aws/resource_aws_s3_bucket_inventory.go

```diff
@@ -0,0 +1,445 @@
+		destinationMap := destinationList[0].(map[string]interface{})
+		bucketList := destinationMap["bucket"].([]interface{})
+		bucketMap := bucketList[0].(map[string]interface{})
+
+		inventoryConfiguration.Destination = &s3.InventoryDestination{
+			S3BucketDestination: expandS3InventoryS3BucketDestination(bucketMap),
+		}
+	}
+
+	input := &s3.PutBucketInventoryConfigurationInput{
+		Bucket: aws.String(bucket),
+		Id:     aws.String(name),
+		InventoryConfiguration: inventoryConfiguration,
+	}
+
+	log.Printf("[DEBUG] Putting S3 bucket inventory configuration: %s", input)
+	err := resource.Retry(1*time.Minute, func() *resource.RetryError {
+		_, err := conn.PutBucketInventoryConfiguration(input)
+		if err != nil {
+			if isAWSErr(err, s3.ErrCodeNoSuchBucket, "") {
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("Error putting S3 bucket inventory configuration: %s", err)
+	}
+
+	d.SetId(fmt.Sprintf("%s:%s", bucket, name))
+
+	return resourceAwsS3BucketInventoryRead(d, meta)
+}
+
+func resourceAwsS3BucketInventoryDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).s3conn
+
+	bucket, name, err := resourceAwsS3BucketInventoryParseID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	input := &s3.DeleteBucketInventoryConfigurationInput{
+		Bucket: aws.String(bucket),
+		Id:     aws.String(name),
+	}
+
+	log.Printf("[DEBUG] Deleting S3 bucket inventory configuration: %s", input)
+	_, err = conn.DeleteBucketInventoryConfiguration(input)
+	if err != nil {
+		if isAWSErr(err, s3.ErrCodeNoSuchBucket, "") || isAWSErr(err, "NoSuchConfiguration", "The specified configuration does not exist.") {
+			return nil
+		}
+		return fmt.Errorf("Error deleting S3 bucket inventory configuration: %s", err)
+	}
+
+	return nil
+}
+
+func resourceAwsS3BucketInventoryRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).s3conn
+
+	bucket, name, err := resourceAwsS3BucketInventoryParseID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	d.Set("bucket", bucket)
+	d.Set("name", name)
+
+	input := &s3.GetBucketInventoryConfigurationInput{
+		Bucket: aws.String(bucket),
+		Id:     aws.String(name),
+	}
+
+	log.Printf("[DEBUG] Reading S3 bucket inventory configuration: %s", input)
+	output, err := conn.GetBucketInventoryConfiguration(input)
+	if err != nil {
+		if isAWSErr(err, s3.ErrCodeNoSuchBucket, "") || isAWSErr(err, "NoSuchConfiguration", "The specified configuration does not exist.") {
+			log.Printf("[WARN] %s S3 bucket inventory configuration not found, removing from state.", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	d.Set("enabled", *output.InventoryConfiguration.IsEnabled)
+	d.Set("included_object_versions", *output.InventoryConfiguration.IncludedObjectVersions)
+	d.Set("optional_fields", flattenStringList(output.InventoryConfiguration.OptionalFields))
```

Comment:
When using `d.Set()` with aggregate types (TypeList, TypeSet, TypeMap), [we should perform error checking to prevent issues where the code is not properly able to set the Terraform state](https://www.terraform.io/docs/extend/best-practices/detecting-drift.html#error-checking-aggregate-types). e.g.

```go
if err := d.Set("optional_fields", flattenStringList(output.InventoryConfiguration.OptionalFields)); err != nil {
  return fmt.Errorf("error setting optional_fields: %s", err)
}
```

Without this check and if there is a type conversion error, Terraform is silently accepting the configuration "as-is" into the Terraform state and unable to detect drift from the API should changes occur outside Terraform.

This applies to `filter`, `schedule`, and `destination` below as well.